### ISSUE TITLE
spriv-opt producing invalid SPIR-V for shader (crashing two drivers tested on)

### ISSUE THREAD

**Issue by danginsburg** [KhronosGroup/SPIRV-Tools#850]
The attached zip file contains three SPIR-V fragment shaders that demonstrate spirv-opt producing invalid SPIR-V.  Unfortunately, the SPIR-V does not fail spirv-val, but it has crashed on two different drivers.  Kerch Holt looked at the SPIR-V and determined there was a problem.  Specifically:

> The following snippet of SPIRV seems incorrect. It doesn’t seem like it is “structured control flow”. The OpSelectionMerge is pointing to %7205 yet there are jumps outside of this region.
> Seems like this code is something like : cond ? goto X : goto Y;  That isn’t allowed in structured flow:
> 1987                OpSelectionMerge %7205 None
> 1988                OpBranchConditional %22516 %21999 %7205
> 1989       %21999 = OpLabel
> 1990                OpBranch %24266
> 1991        %7205 = OpLabel
> 1992                OpBranch %7840
> 1993        %7840 = OpLabel
> 1994       %11181 = OpPhi %float %11180 %21998 %20516 %7205
> 1995       %14351 = OpPhi %v3float %14347 %21998 %7885 %7205
> 1996       %12949 = OpPhi %mat3v4float %15174 %21998 %14541 %7205
> 1997       %16228 = OpPhi %_struct_1268 %15060 %21998 %7120 %7205
> 1998       %17815 = OpFAdd %float %15154 %float_1
> 1999                OpBranch %20259
> 2000       %24266 = OpLabel"

The shaders attached to the bug are:

* generic_vfx_ps_1.spv - this is the SPIR-V that comes out of the HLSL front-end of glslang
* generic.vfx_ps_1.spvopt.spv - this is after running spirv-opt with the following options: spirv-opt.exe --inline-entry-points-exhaustive --convert-local-access-chains --eliminate-local-single-block --eliminate-local-single-store --eliminate-insert-extract --eliminate-dead-code-aggressive --eliminate-dead-branches --merge-blocks --eliminate-local-single-block --eliminate-local-single-store --eliminate-local-multi-store --eliminate-insert-extract --eliminate-dead-code-aggressive
* generic.vfx_ps_1.spvopt_spvremapper.spv - this is the shader we actually send to the driver which runs spirv-remapper

Both of the last two shaders are crashing drivers, so spirv-opt alone is enough to cause the crash.  I included the last one because it has the code that Kerch identified as being incorrect so hopefully makes it easier to identify the issue.

[spirv_opt_bug.zip](https://github.com/KhronosGroup/SPIRV-Tools/files/1342129/spirv_opt_bug.zip)

A secondary issue here is once there is agreement this is invalid spirv, spirv-val should probably get an issue to detect this case since it currently does not.

**Comment by kholtnv** [comment]
The above SPIRV snippet isn't the actual problem (Thanks to Greg F. for pointing that out).
The problem is that the "break" in a switch is interpreted as a "break" from the outer loop instead:
```
497                OpLoopMerge %20205 %7032 None

1501                OpSelectionMerge %7784 None
1502                OpSwitch %int_0 %7784 0 %20555 2 %16948 1 %15148 3 %20270
1503       %20555 = OpLabel
….
1509                OpBranch %20205
1510       %16948 = OpLabel
…..
1525       %21280 = OpFMul %v4float %21562 %13600
1526                OpBranch %20205
1527        %7784 = OpLabel
1528
1529                OpBranch %20205
1530        %7032 = OpLabel
1531                OpBranchConditional %false %24167 %20205
1532       %20205 = OpLabel

```

**Comment by kholtnv** [comment]
Not sure from reading 2.11: Structured Control Flow in SPIRV if this is ill-formed. However, it seems it wouldn't be really possible to even generate using GLSL.  For example, a "break" inside a "switch" would always be interpreted as breaking out of the switch and not an enclosing loop.

**Comment by greg-lunarg** [comment]
(from our private thread)

The OpBranchCondiational %false ... hints to me that in the pre-optimized SPIR-V this switch contained "return" instructions. When we inline such functions, we put it in a loop with the OpBranchCondiational %false ...as the backedge and replace all the "return" instructions with branches to the merge block of the created loop.

Again, like the "break" from the "then" block we discussed earlier in this thread, I think all of the OpBranch %20205 are valid as they are branches to the merge block of an immediately nested loop. Whether they are correct or not is hard to tell until I see the pre-optimized SPIR-V, but I think it is entirely plausible.

While valid, the OpBranchCondiational %false ... construction is a little unusual, and I can certainly imagine a compiler folding this instruction and perhaps losing the fact that this entire construct is a loop. But I am fairly certain the NV compiler has seen this construct in other shaders.

Bottom line: I think this is still valid (and possibly even correct) code :)

**Comment by dneto0** [comment]
I agree with Greg.  This is what the inliner has to do when inlining a function that has multiple returns inside a selection construct.   I believe it's valid code.  Details below.

General note on debugging control flow issues:  spirv-cfg is really useful for seeing the control flow graph.

The CFG for generic.vfx_ps_1.spvopt_spvremapper.spv is:

![r](https://user-images.githubusercontent.com/13054574/31026811-c1fbf242-a515-11e7-9eb1-3e2bd05ddb27.png)


The interesing bit is on the bottom right of the graph.

The branches from the switch cases out to %20205 are valid:
-  The loop starts at %6607, and names %20205 as its merge block
-  Block 7784 is the merge block for the selection rooted at block 25066.
-  The branches from within the switch from blocks 20270  20555 16948 15148 out to the loop's merge block 20205 are valid because that makes those blocks "break blocks":

> Break Block: A block containing a branch to the Merge Block of a loop header’s merge instruction.

And from 2.11:

> the only blocks in a construct that can branch outside the construct are ... (one case being) a break block for the innermost loop it is nested inside of.

**Comment by dneto0** [comment]
Incidentally, I also verified the OpPhi instructions in block 20205, and it looks right.

Here's a smaller test case showing what's going on:

```
#version 310 es

int a;

void bar(int s) {
  switch(s) {
    case 0: a = 0; return;
    case 1: a = 1; return;
    default:
     break;
  }
  a = 2;
}

void main() {
  bar(12);
}
```
Compiles directly to:
```
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Fragment %main "main"
               OpExecutionMode %main OriginUpperLeft
               OpSource ESSL 310
               OpName %main "main"
               OpName %bar_i1_ "bar(i1;"
               OpName %s "s"
               OpName %a "a"
               OpName %param "param"
               OpDecorate %s RelaxedPrecision
               OpDecorate %12 RelaxedPrecision
               OpDecorate %a RelaxedPrecision
       %void = OpTypeVoid
          %3 = OpTypeFunction %void
        %int = OpTypeInt 32 1
%_ptr_Function_int = OpTypePointer Function %int
          %8 = OpTypeFunction %void %_ptr_Function_int
%_ptr_Private_int = OpTypePointer Private %int
          %a = OpVariable %_ptr_Private_int Private
      %int_0 = OpConstant %int 0
      %int_1 = OpConstant %int 1
      %int_2 = OpConstant %int 2
     %int_12 = OpConstant %int 12
       %main = OpFunction %void None %3
          %5 = OpLabel
      %param = OpVariable %_ptr_Function_int Function
               OpStore %param %int_12
         %28 = OpFunctionCall %void %bar_i1_ %param
               OpReturn
               OpFunctionEnd
    %bar_i1_ = OpFunction %void None %8
          %s = OpFunctionParameter %_ptr_Function_int
         %11 = OpLabel
         %12 = OpLoad %int %s
               OpSelectionMerge %16 None
               OpSwitch %12 %15 0 %13 1 %14
         %15 = OpLabel
               OpBranch %16
         %13 = OpLabel
               OpStore %a %int_0
               OpReturn
         %14 = OpLabel
               OpStore %a %int_1
               OpReturn
         %16 = OpLabel
               OpStore %a %int_2
               OpReturn
               OpFunctionEnd
```
Control flow graph:

![f](https://user-images.githubusercontent.com/13054574/31027554-883b3c0e-a518-11e7-80de-933472aa257c.png)

Running spirv-opt --inline-entry-points-exhaustive --eliminate-dead-functions on it yields:

```
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Fragment %main "main"
               OpExecutionMode %main OriginUpperLeft
               OpSource ESSL 310
               OpName %main "main"
               OpName %a "a"
               OpName %param "param"
               OpDecorate %a RelaxedPrecision
       %void = OpTypeVoid
          %3 = OpTypeFunction %void
        %int = OpTypeInt 32 1
%_ptr_Function_int = OpTypePointer Function %int
          %8 = OpTypeFunction %void %_ptr_Function_int
%_ptr_Private_int = OpTypePointer Private %int
          %a = OpVariable %_ptr_Private_int Private
      %int_0 = OpConstant %int 0
      %int_1 = OpConstant %int 1
      %int_2 = OpConstant %int 2
     %int_12 = OpConstant %int 12
       %bool = OpTypeBool
      %false = OpConstantFalse %bool
       %main = OpFunction %void None %3
          %5 = OpLabel
      %param = OpVariable %_ptr_Function_int Function
               OpStore %param %int_12
               OpBranch %29
         %29 = OpLabel
               OpLoopMerge %30 %31 None
               OpBranch %32
         %32 = OpLabel
         %33 = OpLoad %int %param
               OpSelectionMerge %34 None
               OpSwitch %33 %35 0 %36 1 %37
         %35 = OpLabel
               OpBranch %34
         %36 = OpLabel
               OpStore %a %int_0
               OpBranch %30
         %37 = OpLabel
               OpStore %a %int_1
               OpBranch %30
         %34 = OpLabel
               OpStore %a %int_2
               OpBranch %30
         %31 = OpLabel
               OpBranchConditional %false %29 %30
         %30 = OpLabel
               OpReturn
               OpFunctionEnd
```
with control flow graph:

![f inline](https://user-images.githubusercontent.com/13054574/31027603-c6b59a4c-a518-11e7-87dc-d6dc9efd7f9b.png)

**Comment by dneto0** [comment]
I checked, and I believe the Vulkan CTS does not cover this case.

It ought to be covered in createLoopTests:  https://github.com/KhronosGroup/VK-GL-CTS/blob/master/external/vulkancts/modules/vulkan/spirv_assembly/vktSpvAsmInstructionTests.cpp#L6575

**Comment by dneto0** [comment]
@danginsburg : More to the point, to work around this problem it would be sufficient if your helper functions avoid returning from within a selection construct.

The inliner is (currently) unable to inline a function that returns from inside a loop construct. 
https://github.com/KhronosGroup/SPIRV-Tools/blob/master/source/opt/inline_pass.cpp#L658 
We haven't yet taught the optimizer how to transform such a function  so that it returns only once, at the end.  It's possible, but would require much more machinery.

**Comment by jekstrand** [comment]
I just got informed about this via a patch from @BNieuwenhuizen to mesa-dev:

https://lists.freedesktop.org/archives/mesa-dev/2017-October/171254.html

I think I'm fine with allowing loop breaks inside a switch even if it isn't something that's allowed in the source language.  However, we really need tests and this is probably something SPIR-V opt should avoid for now since there are no CTS tests and we (mesa) may not the only ones to have the bug.

**Comment by danginsburg** [comment]
Confirmed @BNieuwenhuizen patch to mesa fixes the crash on RADV and renders correctly.  I tested the shader on on AMD GPUPRO and it worked fine.  So looks like NVIDIA is the last one crashing.  Agreed about need for a conformance test.

**Comment by danginsburg** [comment]
I thought it was also worth tracking down the HLSL code that was the root of the problem.  It seems to me there might be a better code generation pattern we should maybe pursue.

So a snippet of the actual HLSL that causes the problem:
```
float4 SampleWeightBasedUponDistancesToSamplePointInTexels( float4 vecXDifferences, float4 vecYDifferences, int nFilterType )
{
	switch( nFilterType )
	{
		case SHADOWFILTER_GAUSSIAN:
		{
			float4 flDistance = ( vecXDifferences * vecXDifferences + vecYDifferences * vecYDifferences );
			return exp( -1.4 * flDistance );
		}
		
		case SHADOWFILTER_BOX:
		{
			return float4( 1.0, 1.0, 1.0, 1.0 );
		}
		
		case SHADOWFILTER_TRIANGLE:
		{
			return max( float4( 0.0, 0.0, 0.0, 0 ), ( 4 - vecXDifferences ) * ( 4 - vecYDifferences ) );
		}

		case SHADOWFILTER_SINGLE_TAP_PCF:
		{
			return ( 1.0 - vecXDifferences ) * ( 1.0 - vecYDifferences );
		}
	}

	return float4( 0.0, 0.0, 0.0, 0.0 );
}

// Called in a loop later on:
LightingTerms_t ComputeForwardLighting( float4 vPosition4Cs, float3 vPositionWs, float3 vNormalWs, float3 vGeometricNormalWs, float flRoughness, float flSsaoOcclusionFactor,
										float flReflectance, float flFresnelExponent )
{
		// ...
		[loop] for ( float ss = 0.0; ss < flNumShadowedSpotsToDraw; ss += 1.0 )
		{
			//..
			float flShadowScalar = SampleShadowMap( spotLight, vClipSpace.xy, spotLight.vShadowMapBiasAndScale, vPositionWs.xyz, vPositionToLightDirWs.xyz, SHADOWFILTER_GAUSSIAN, vGeometricNormalWs, PassToArgTexture2D( g_tLightDepthBuffer ) );
			//..
		}
		//..
}
```
If I replace SampleWeightBasedUponDistancesToSamplePointsInTexels with the return outside the case statements then it no longer crashes on NVIDIA:
```
loat4 SampleWeightBasedUponDistancesToSamplePointInTexels( float4 vecXDifferences, float4 vecYDifferences, int nFilterType )
{
	float vResult = float4( 0.0, 0.0, 0.0, 0.0 );
	switch( nFilterType )
	{
		case SHADOWFILTER_GAUSSIAN:
		{
			float4 flDistance = ( vecXDifferences * vecXDifferences + vecYDifferences * vecYDifferences );
			vResult = exp( -1.4 * flDistance );
		}
		
		case SHADOWFILTER_BOX:
		{
			vResult = float4( 1.0, 1.0, 1.0, 1.0 );
		}
		
		case SHADOWFILTER_TRIANGLE:
		{
			vResult = max( float4( 0.0, 0.0, 0.0, 0 ), ( 4 - vecXDifferences ) * ( 4 - vecYDifferences ) );
		}

		case SHADOWFILTER_SINGLE_TAP_PCF:
		{
			vResult = ( 1.0 - vecXDifferences ) * ( 1.0 - vecYDifferences );
		}
	}

	return vResult;

}
```
I would note also that if you look at the call to SampleWeightBasedUponDistancesToSamplePointsInTexels the nFilterType is hard-coded to SHADOWFILTER_GAUSSIAN so with constant propagation I would have thought it could avoid evaluation of the other branches in the switch statement.

I'm not sure what the action should be here, but it seems like we should be able to generate much nicer SPIR-V code here either by propagating the nFilterType constant or maybe some transformation on SampleWeightBasedUponDistancesToSamplePointsInTexels  so that the SPIR-V generated looks more like the second version.

**Comment by greg-lunarg** [comment]
Thanks Dan.

Just a quick technical observation: I think you want a "break" after each case in your rewrite to better reflect the original logic, but I think even then you would see the NV crash go away.

I think having spirv-opt (or glslang or some other post-processor) do the rewrite you did would be possible, but it would start getting more complicated the more general we made it, for example, what if some cases returned and others didn't, or what if some of the cases had their returns nested deeper in control flow. I think it would start to get into the kind of complex rewrite that the SPIR-V WG said that they did not want spirv-opt to generate.

Your good observation on the hard-coding of nFilterType I think indicates a better solution. Simple constant folding, upgrading the dead branch elimination to deal with switches and trivial loop elimination would fix this problem and is likely the way that fxc deals with this.

In general, I am guessing that fxc doesn't do anything overly complex or inefficient, so I think requirements on spirv-opt in dealing with HLSL will be similar.

My recommendation would be to pursue the approach outlined above (constant folding, etc). In the meantime, we could (temporarily) disable inlining of functions with early returns in functions. We already do this for functions with early returns in loops, so adding switches would be fairly easy.

**Comment by greg-lunarg** [comment]
Small correction: constant folding is not required here. Just expanding dead branch to deal with constant switch and trivial loop elimination.

**Comment by greg-lunarg** [comment]
Another small correction: we really don't need to get rid of the loop (although it would be nice). Since the problem is breaking out of a loop from switch, eliminating the switch should be sufficient.

**Comment by dneto0** [comment]
Good discussion.

It turns out that I can generate the "bad" control flow graph with plain GLSL. Let's revisit the small example from my comment https://github.com/KhronosGroup/SPIRV-Tools/issues/850#issuecomment-333186062    
But instead of escaping the switch construct with a return, I can equally escape it with a "continue":

```
#version 450

int a;
int s;

void main() {
  bool keep_going = true;
  for (; keep_going; keep_going = false) {
    switch (s) {
      case 0:
        a = 0;
        continue;
      case 1:
        a = 1;
        continue;
      default:
        continue;
    }
  }
  a = 2;
}
```
That produces:
```
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Vertex %main "main"
               OpSource GLSL 450
               OpName %main "main"
               OpName %keep_going "keep_going"
               OpName %s "s"
               OpName %a "a"
       %void = OpTypeVoid
          %3 = OpTypeFunction %void
       %bool = OpTypeBool
%_ptr_Function_bool = OpTypePointer Function %bool
       %true = OpConstantTrue %bool
        %int = OpTypeInt 32 1
%_ptr_Private_int = OpTypePointer Private %int
          %s = OpVariable %_ptr_Private_int Private
          %a = OpVariable %_ptr_Private_int Private
      %int_0 = OpConstant %int 0
      %int_1 = OpConstant %int 1
      %false = OpConstantFalse %bool
      %int_2 = OpConstant %int 2
       %main = OpFunction %void None %3
          %5 = OpLabel
 %keep_going = OpVariable %_ptr_Function_bool Function
               OpStore %keep_going %true
               OpBranch %10
         %10 = OpLabel
               OpLoopMerge %12 %13 None
               OpBranch %14
         %14 = OpLabel
         %15 = OpLoad %bool %keep_going
               OpBranchConditional %15 %11 %12
         %11 = OpLabel
         %19 = OpLoad %int %s
               OpSelectionMerge %23 None
               OpSwitch %19 %22 0 %20 1 %21
         %22 = OpLabel
               OpBranch %13
         %20 = OpLabel
               OpStore %a %int_0
               OpBranch %13
         %21 = OpLabel
               OpStore %a %int_1
               OpBranch %13
         %23 = OpLabel
               OpBranch %13
         %13 = OpLabel
               OpStore %keep_going %false
               OpBranch %10
         %12 = OpLabel
               OpStore %a %int_2
               OpReturn
               OpFunctionEnd

```
with CFG:
![x](https://user-images.githubusercontent.com/13054574/31230674-b19970c2-a9b2-11e7-8814-b38b67e11c51.png)

This is the same pattern:
- Selection header is 11
- Selection merge block is 23
- Blocks 20, 21, 22 are selection cases constructs but break out to the continue target 13 of the enclosing loop.

Backing up to @danginsburg 's rewrite of the original HLSL (and corrections of course).   It's a good illustration of a general pattern  should always be doable, either in original source or in SPIR-V.
The pattern is:  write the result to a new function-scope variable, then rewire the control flow to always fall out of the bottom of the function.  That's what a proposed fix in the SPIR-V domain would look like.

That would also handle the case of multiple returns in a loop.  It's just that @greg-lunarg nor anybody else has invested in writing that solution.


I understand there's a gap between being correct and being helpful.  Drivers can be fixed only so quickly, so in the meantime I offered the mitigation strategy (don't return early) at the higher level language.  I'll keep this issue open for workaround in the SPIR-V domain.

Given that I can generate the troublesome CFG with plain GLSL I'd recommend a pass that transforms a function to have a single return in an exit block that post-dominates all other blocks (modulo unreachable and OpKill). This would be equivalent to LLVM's 'mergereturn' transform: http://llvm.org/doxygen/UnifyFunctionExitNodes_8cpp_source.html 
Then this mergereturn pass could be used as a preprocessing step or its guts could be used by the inliner to avoid the troublesome graphs.

Proposed plan of action:
- drivers fix their implementations
- make Vulkan CTS tests.  I already filed Vulkan CTS issue 722 (inside Khronos walls)
- In SPIRV-Tools
  - Write a merge-return pass
  - Update inliner to use merge-return internals to avoid the troublesome CFGs


Last notes:
- Constant propagation and other optimizations would be very nice to have.  But that's orthogonal to the root cause of this issue
- Microsoft does not require DXIL to have structured control flow (see https://github.com/Microsoft/DirectXShaderCompiler/blob/master/docs/DXIL.rst#control-flow-restrictions), so comparisons with fxc (and dxc) can't be definitive about what we ought to do in SPIR-V for Vulkan.

**Comment by dneto0** [comment]
I filed #857 for making a merge-return pass

**Comment by greg-lunarg** [comment]
I am currently adapting dead-branch-elimination to handle switches with constant switch values.

I realize this doesn't fix the more general problem of returns in switches. For example, it would not help if the switch value here is not constant.

But beefing up dead-branch-elimination will fix this issue and is probably the shortest path to doing so. It will also be useful going forward as part of the ongoing size reduction and legalization effort. fxc clearly does dead case elimination and it is more than likely that such a capability will be required as part of legalization.

**Comment by greg-lunarg** [comment]
PR #863 submitted for consideration.

**Comment by dneto0** [comment]
#863 has been merged

I believe that should fix this particular case of the troublesome CFG pattern.  

I re-ran the sequence of passes you listed (except spirv-remap, and adding --eliminate-dead-functions just for clarity), and the resulting control flow graph is:

![foo](https://user-images.githubusercontent.com/13054574/31506224-d4c5fefc-af44-11e7-8eda-7f773aaaa36a.png)

The corresponding area is at the loop headed at block 6607 toward the bottom right.  The single block at 6610 is where previously we had the switch construct that escaped out to the the loop's merge block.  Now 6610 is just a single block and it branches to the loop's merge block.   This is valid and I expect it to work in your drivers.

I'll close for now.  Please reopen or file another issue if the problem perisists.

I still have the #857 task open, to make a merge-return pass.  The merge-return pass could be used address more general cases where the troublesome CFG can occur, and to enable the inliner to work in all cases.

**Comment by danginsburg** [comment]
Just wanted to confirm that #863 fixed the crash for this case on NVIDIA.  Thanks all.
